src/dbnode/persist/fs/migration/migration.go

```diff
@@ -0,0 +1,108 @@
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package migration
+
+import (
+	"github.com/m3db/m3/src/dbnode/namespace"
+	"github.com/m3db/m3/src/dbnode/persist"
+	"github.com/m3db/m3/src/dbnode/persist/fs"
+	"github.com/m3db/m3/src/dbnode/storage"
+	xtime "github.com/m3db/m3/src/x/time"
+)
+
+// Task interface is implemented by tasks that wish to perform a data migration
+// on a fileset. This typically involves updating files in a fileset that were created by
+// a previous version of the database client.
+type Task interface {
+	// Run is the set of steps to successfully complete a migration.
+	Run() error
+}
+
+// toVersion1_1Task is an object responsible for migrating a fileset to version 1.1.
+type toVersion1_1Task struct {
+	newMergerFn       fs.NewMergerFn
+	infoFileResult    fs.ReadInfoFileResult
+	shard             uint32
+	namespaceMetadata namespace.Metadata
+	persistManager    persist.Manager
+	sOpts             storage.Options
+	fsOpts            fs.Options
+}
+
+// ShouldMigrateToVersion1_1 indicates whether the fileset should be migrated to 1.1 or not.
+func ShouldMigrateToVersion1_1(info fs.ReadInfoFileResult) bool {
```

Comment:
More of a question here, but could this be flipped as `func MigrationTask(info fs.ReadInfoFileResult) (Task, bool)` which would return the version to migrate to and a bool indicating if we should migrate? This could represent significant refactoring though so don't mind too much; could also be good future work if we have a `toVersion1_2Task` in future

Reply:
I like that idea. Doesn't end up being much of a refactor at all to do. I'll add that.